File: src-tauri/src/ssh/connection_manager.rs
//! SSH connection lifecycle over russh 0.60. Reconciled against the Go oracle
//! `internal/ssh/connection_manager.go` (State guards ONLY the conn pointer; the
//! mutex is never held across I/O — old conn dropped after the lock is released).
//!
//! russh 0.48 → 0.60 adaptations:
//!   - keys moved: `russh_keys::*` → `russh::keys::*` (no separate crate).
//!   - `client::Handler` is a native async trait (no `#[async_trait]`);
//!     `check_server_key(&mut self, &ssh_key::PublicKey) -> Result<bool, Error>`.
//!   - pubkey auth takes `PrivateKeyWithHashAlg::new(Arc<PrivateKey>, hash_alg)`.
//!   - `authenticate_*` return `AuthResult`; success via `.success()`.
//!   - agent auth: `authenticate_publickey_with(user, PublicKey, hash_alg, &mut AgentClient)`.

use std::borrow::Cow;
use std::sync::{Arc, Mutex};

use russh::client;
use russh::keys::agent::client::AgentClient;
use russh::keys::ssh_key::{Algorithm, EcdsaCurve, HashAlg, PublicKey};
use russh::keys::{load_secret_key, PrivateKeyWithHashAlg};
use russh::Preferred;
use russh_sftp::client::SftpSession;

use super::agent_discovery;
use super::config_parser;
use super::known_hosts::{
    check_or_learn, default_known_hosts_path, is_algorithm_allowed, Decision, DecisionResult,
};
use super::retry::Dialer;
use super::types::{ConfigHostEntry, ConnectionConfig, ConnectionStatus};

// Host-key allowlist for negotiation (no ssh-rsa SHA-1, no DSS).
const SAFE_HOST_KEY_ALGOS: &[Algorithm] = &[
    Algorithm::Ed25519,
    Algorithm::Ecdsa { curve: EcdsaCurve::NistP256 },
    Algorithm::Ecdsa { curve: EcdsaCurve::NistP384 },
    Algorithm::Ecdsa { curve: EcdsaCurve::NistP521 },
    Algorithm::Rsa { hash: Some(HashAlg::Sha512) },
    Algorithm::Rsa { hash: Some(HashAlg::Sha256) },
];

fn safe_preferred() -> Preferred {
    Preferred {
        key: Cow::Borrowed(SAFE_HOST_KEY_ALGOS),
        ..Preferred::DEFAULT
    }
}

// ─── host-key handler (the security boundary) ────────────────────────────────

struct SshHandler {
    host: String,
    port: u16,
    decision: Arc<Mutex<Option<DecisionResult>>>,
}

impl SshHandler {
    fn record(&self, decision: DecisionResult) {
        if let Ok(mut slot) = self.decision.lock() {
            *slot = Some(decision);
        }
    }
}

impl client::Handler for SshHandler {
    type Error = russh::Error;

    /// Accept ONLY Trusted/Learned. Reject (`Ok(false)`, NEVER `Ok(true)`) on an
    /// off-allowlist algorithm or a changed key; fail closed (`Ok(false)`) on any
    /// error. The stored decision lets `connect` surface a descriptive message.
    async fn check_server_key(
        &mut self,
        server_public_key: &PublicKey,
    ) -> Result<bool, Self::Error> {
        // Boundary #1: algorithm allowlist (blocks ssh-rsa SHA-1 / DSS).
        let algo = server_public_key.algorithm().as_str().to_string();
        if !is_algorithm_allowed(&algo) {
            self.record(DecisionResult::algorithm_rejected(algo));
            return Ok(false);
        }
        // Boundary #2: managed known_hosts TOFU + reject-on-change.
        let path = match default_known_hosts_path() {
            Ok(p) => p,
            Err(_) => return Ok(false),
        };
        let result = match check_or_learn(&path, &self.host, self.port, server_public_key) {
            Ok(r) => r,
            Err(_) => return Ok(false),
        };
        let accept = matches!(result.kind, Decision::TrustedExisting | Decision::LearnedNew);
        self.record(result);
        Ok(accept)
    }
}

// ─── connection + state ──────────────────────────────────────────────────────

pub struct Connection {
    _handle: client::Handle<SshHandler>,
    _sftp: SftpSession,
    pub remote_projects_path: String,
    _remote_todos_path: String,
    host: String,
}

/// Mirrors Go `State` — the mutex guards ONLY the conn pointer, never held
/// across I/O; the old connection is dropped after the lock is released.
#[derive(Default)]
pub struct State {
    conn: Mutex<Option<Connection>>,
}

impl State {
    pub fn new() -> Self {
        Self::default()
    }

    /// Mirrors Go `GetStatus`.
    pub fn get_status(&self) -> ConnectionStatus {
        let guard = self.conn.lock().unwrap_or_else(|e| e.into_inner());
        match &*guard {
            Some(c) => ConnectionStatus {
                state: "connected".to_string(),
                host: Some(c.host.clone()),
                error: None,
                remote_projects_path: Some(c.remote_projects_path.clone()),
                retry_attempt: None,
                max_retries: None,
            },
            None => ConnectionStatus::disconnected(),
        }
    }

    /// Mirrors Go `SetConn` — swap under lock, drop the old conn after release.
    pub fn set_conn(&self, conn: Option<Connection>) {
        let old = {
            let mut guard = self.conn.lock().unwrap_or_else(|e| e.into_inner());
            std::mem::replace(&mut *guard, conn)
        };
        drop(old);
    }

    /// Mirrors Go `ClearConn`.
    pub fn clear_conn(&self) {
        self.set_conn(None);
    }
}

// ─── dialer + entry points ───────────────────────────────────────────────────

/// Production `Dialer` (mirrors Go `NetDialer`). russh does its own TCP dial.
pub struct NetDialer;

impl Dialer for NetDialer {
    type Conn = Connection;
    async fn connect(&self, config: &ConnectionConfig) -> Result<Connection, String> {
        connect(config).await
    }
}

/// Mirrors Go `TestConnection` — connect via the dialer, then drop.
pub async fn test_connection<D: Dialer>(config: &ConnectionConfig, dialer: &D) -> Result<(), String> {
    let _conn = dialer.connect(config).await?;
    Ok(())
}

fn resolve_target(
    config: &ConnectionConfig,
    ssh_config: &Option<ConfigHostEntry>,
) -> (String, u16, String) {
    let host = ssh_config
        .as_ref()
        .and_then(|c| c.host_name.clone())
        .unwrap_or_else(|| config.host.clone());
    let port = if config.port != 22 {
        config.port
    } else {
        ssh_config.as_ref().and_then(|c| c.port).unwrap_or(config.port)
    };
    let username = if config.username.is_empty() {
        ssh_config
            .as_ref()
            .and_then(|c| c.user.clone())
            .unwrap_or_else(|| std::env::var("USER").unwrap_or_else(|_| "root".to_string()))
    } else {
        config.username.clone()
    };
    (host, port, username)
}

fn enforce_host_key_decision(
    decision: Option<DecisionResult>,
    host: &str,
    port: u16,
) -> Result<(), String> {
    if let Some(d) = decision {
        match d.kind {
            Decision::KeyChanged => {
                return Err(format!(
                    "SSH host key changed for {host}:{port} — possible MITM. \
                     recorded={} offered={}. \
                     Edit ~/.claude/ssh/known_hosts manually to recover.",
                    d.recorded_fingerprint, d.offered_fingerprint
                ));
            }
            Decision::AlgorithmRejected => {
                return Err(format!(
                    "SSH host key algorithm {} is not on the allowlist",
                    d.algorithm
                ));
            }
            _ => {}
        }
    }
    Ok(())
}

/// Mirrors Go `Connect` — dial, verify host key, authenticate, open SFTP,
/// resolve the remote projects path.
pub async fn connect(config: &ConnectionConfig) -> Result<Connection, String> {
    let ssh_config = config_parser::resolve_host(&config.host);
    let (actual_host, actual_port, username) = resolve_target(config, &ssh_config);

    let russh_config = Arc::new(client::Config {
        inactivity_timeout: Some(std::time::Duration::from_secs(120)),
        preferred: safe_preferred(),
        ..Default::default()
    });

    let decision_slot = Arc::new(Mutex::new(None));
    let handler = SshHandler {
        host: actual_host.clone(),
        port: actual_port,
        decision: Arc::clone(&decision_slot),
    };

    let connect_result =
        client::connect(russh_config, (actual_host.as_str(), actual_port), handler).await;

    // Surface the security decision first (a rejected key aborts the handshake,
    // so connect_result is Err — but we override with a descriptive message).
    let decision = decision_slot
        .lock()
        .ok()
        .and_then(|g| g.clone());
    enforce_host_key_decision(decision, &actual_host, actual_port)?;

    let mut session = connect_result.map_err(|e| format!("SSH connection failed: {e}"))?;

    authenticate(&mut session, config, &ssh_config, &username)
        .await
        .map_err(|e| format!("SSH authentication failed: {e}"))?;

    let channel = session
        .channel_open_session()
        .await
        .map_err(|e| format!("Failed to open SSH channel: {e}"))?;
    channel
        .request_subsystem(true, "sftp")
        .await
        .map_err(|e| format!("Failed to request SFTP subsystem: {e}"))?;
    let sftp = SftpSession::new(channel.into_stream())
        .await
        .map_err(|e| format!("Failed to start SFTP session: {e}"))?;

    let remote_home = resolve_remote_home(&session).await;
    let remote_projects_path =
        find_remote_projects_path(&sftp, &username, remote_home.as_deref()).await;
    let remote_todos_path = match &remote_home {
        Some(home) => format!("{home}/.claude/todos"),
        None => format!("/home/{username}/.claude/todos"),
    };

    Ok(Connection {
        _handle: session,
        _sftp: sftp,
        remote_projects_path,
        _remote_todos_path: remote_todos_path,
        host: config.host.clone(),
    })
}

// ─── authentication ──────────────────────────────────────────────────────────

async fn authenticate(
    session: &mut client::Handle<SshHandler>,
    config: &ConnectionConfig,
    ssh_config: &Option<ConfigHostEntry>,
    username: &str,
) -> Result<(), String> {
    match config.auth_method.as_str() {
        "password" => auth_password(session, config, username).await,
        "privateKey" => auth_private_key(session, config, username).await,
        "agent" => auth_agent(session, username).await,
        _ => auth_auto(session, ssh_config, username).await,
    }
}

async fn auth_password(
    session: &mut client::Handle<SshHandler>,
    config: &ConnectionConfig,
    username: &str,
) -> Result<(), String> {
    let password = config.password.as_deref().ok_or("Password required")?;
    let ok = session
        .authenticate_password(username, password)
        .await
        .map_err(|e| e.to_string())?
        .success();
    if ok {
        Ok(())
    } else {
        Err("Password authentication failed".to_string())
    }
}

async fn auth_private_key(
    session: &mut client::Handle<SshHandler>,
    config: &ConnectionConfig,
    username: &str,
) -> Result<(), String> {
    let key_path = config.private_key_path.as_deref().unwrap_or("~/.ssh/id_rsa");
    let expanded = expand_tilde(key_path);
    let key = load_secret_key(&expanded, None)
        .map_err(|e| format!("Cannot read private key at {expanded}: {e}"))?;
    let pk = PrivateKeyWithHashAlg::new(Arc::new(key), None);
    let ok = session
        .authenticate_publickey(username, pk)
        .await
        .map_err(|e| e.to_string())?
        .success();
    if ok {
        Ok(())
    } else {
        Err("Public key authentication failed".to_string())
    }
}

async fn auth_agent(
    session: &mut client::Handle<SshHandler>,
    username: &str,
) -> Result<(), String> {
    let socket = agent_discovery::discover_agent_socket()
        .await
        .ok_or("SSH agent socket not found")?;
    let mut agent = AgentClient::connect_uds(&socket)
        .await
        .map_err(|e| format!("Cannot connect to SSH agent: {e}"))?;
    let identities = agent
        .request_identities()
        .await
        .map_err(|e| format!("Failed to list agent identities: {e}"))?;

    for identity in &identities {
        let pubkey = identity.public_key().into_owned();
        if let Ok(result) = session
            .authenticate_publickey_with(username, pubkey, None, &mut agent)
            .await
        {
            if result.success() {
                return Ok(());
            }
        }
    }
    Err(format!(
        "Agent authentication failed ({} identities tried)",
        identities.len()
    ))
}

async fn auth_auto(
    session: &mut client::Handle<SshHandler>,
    ssh_config: &Option<ConfigHostEntry>,
    username: &str,
) -> Result<(), String> {
    // 1. Config identity file → default key files.
    if ssh_config.as_ref().map(|c| c.has_identity_file).unwrap_or(false)
        && try_default_keys(session, username).await
    {
        return Ok(());
    }

    // 2. SSH agent.
    if let Ok(mut agent) = AgentClient::connect_env().await {
        if let Ok(identities) = agent.request_identities().await {
            for identity in &identities {
                let pubkey = identity.public_key().into_owned();
                if let Ok(result) = session
                    .authenticate_publickey_with(username, pubkey, None, &mut agent)
                    .await
                {
                    if result.success() {
                        return Ok(());
                    }
                }
            }
        }
    }

    // 3. Default key files.
    if try_default_keys(session, username).await {
        return Ok(());
    }

    Err("No authentication method succeeded".to_string())
}

async fn try_default_keys(session: &mut client::Handle<SshHandler>, username: &str) -> bool {
    for name in ["id_ed25519", "id_rsa", "id_ecdsa"] {
        let path = match dirs::home_dir() {
            Some(h) => h.join(".ssh").join(name),
            None => continue,
        };
        if let Ok(key) = load_secret_key(&path, None) {
            let pk = PrivateKeyWithHashAlg::new(Arc::new(key), None);
            if let Ok(result) = session.authenticate_publickey(username, pk).await {
                if result.success() {
                    return true;
                }
            }
        }
    }
    false
}

// ─── remote path resolution ──────────────────────────────────────────────────

async fn resolve_remote_home(session: &client::Handle<SshHandler>) -> Option<String> {
    let channel = session.channel_open_session().await.ok()?;
    channel.exec(true, "printf %s \"$HOME\"").await.ok()?;

    let mut stdout = String::new();
    let mut stream = channel.into_stream();
    let mut buf = vec![0u8; 4096];
    loop {
        match tokio::io::AsyncReadExt::read(&mut stream, &mut buf).await {
            Ok(0) => break,
            Ok(n) => stdout.push_str(&String::from_utf8_lossy(&buf[..n])),
            Err(_) => break,
        }
    }

    let trimmed = stdout.trim().to_string();
    if trimmed.starts_with('/') {
        Some(trimmed)
    } else {
        None
    }
}

async fn find_remote_projects_path(
    sftp: &SftpSession,
    username: &str,
    remote_home: Option<&str>,
) -> String {
    let mut candidates = Vec::new();
    if let Some(home) = remote_home {
        candidates.push(format!("{home}/.claude/projects"));
    }
    candidates.push(format!("/home/{username}/.claude/projects"));
    candidates.push(format!("/Users/{username}/.claude/projects"));
    candidates.push("/root/.claude/projects".to_string());
    candidates.dedup();

    for candidate in &candidates {
        if sftp.try_exists(candidate.as_str()).await.unwrap_or(false) {
            return candidate.clone();
        }
    }

    match remote_home {
        Some(home) => format!("{home}/.claude/projects"),
        None => format!("/home/{username}/.claude/projects"),
    }
}

fn expand_tilde(path: &str) -> String {
    if let Some(rest) = path.strip_prefix('~') {
        if let Some(home) = dirs::home_dir() {
            return format!("{}{rest}", home.display());
        }
    }
    path.to_string()
}

File: internal/ssh/connection_manager.go
// Package ssh — connection_manager.go ports src-tauri/src/ssh/connection_manager.rs.
//
// Concurrency contract (arch H4):
//   - s.mu guards ONLY the in-memory conn pointer swap.
//   - s.mu is acquired, conn is read/swapped, then s.mu is released BEFORE any
//     network I/O (dial, auth, SFTP). Network calls are never made while holding
//     the mutex. This prevents the "mutex held across await" anti-pattern from
//     the original Rust tokio::sync::Mutex usage.
//
// The Dialer interface isolates network calls so tests inject a fake.
package ssh

import (
	"fmt"
	"net"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	gossh "golang.org/x/crypto/ssh"
	"golang.org/x/crypto/ssh/agent"
	"github.com/pkg/sftp"
)

// Dialer is the network abstraction injected by callers (real or fake in tests).
// Dial opens a TCP connection to addr. In production: net.Dial("tcp", addr).
type Dialer interface {
	Dial(network, addr string) (net.Conn, error)
}

// NetDialer is the production Dialer.
type NetDialer struct{}

func (NetDialer) Dial(network, addr string) (net.Conn, error) {
	return net.DialTimeout(network, addr, 30*time.Second)
}

// Connection holds the live SSH+SFTP session.
type Connection struct {
	client             *gossh.Client
	sftp               *sftp.Client
	RemoteProjectsPath string
	remoteTodosPath    string
	host               string
}

// close tears down the SFTP client and the SSH client.
func (c *Connection) close() {
	if c.sftp != nil {
		_ = c.sftp.Close()
	}
	if c.client != nil {
		_ = c.client.Close()
	}
}

// State mirrors SshState — holds the current (optional) live Connection.
// The mutex only guards the conn pointer; it is never held across I/O.
type State struct {
	mu   sync.Mutex
	conn *Connection
}

// GetStatus mirrors SshState::get_status — snapshot under lock.
func (s *State) GetStatus() ConnectionStatus {
	s.mu.Lock()
	conn := s.conn
	s.mu.Unlock()

	if conn == nil {
		return Disconnected()
	}
	rp := conn.RemoteProjectsPath
	h := conn.host
	return ConnectionStatus{
		State:              "connected",
		Host:               &h,
		RemoteProjectsPath: &rp,
	}
}

// SetConn replaces the live connection under lock. Previous connection is
// closed after the lock is released (closing is I/O, so must happen outside).
func (s *State) SetConn(c *Connection) {
	s.mu.Lock()
	old := s.conn
	s.conn = c
	s.mu.Unlock()

	if old != nil {
		old.close()
	}
}

// ClearConn sets the connection to nil and closes the old one.
func (s *State) ClearConn() {
	s.SetConn(nil)
}

// Connect dials and authenticates; returns a live Connection.
// The dialer is injected so tests can supply a fake.
// NEVER called while holding s.mu.
func Connect(cfg *ConnectionConfig, dialer Dialer) (*Connection, error) {
	// Resolve SSH config overrides.
	entry := ResolveHost(cfg.Host)

	actualHost := cfg.Host
	if entry != nil && entry.HostName != nil {
		actualHost = *entry.HostName
	}

	actualPort := cfg.Port
	if actualPort == 22 && entry != nil && entry.Port != nil {
		actualPort = *entry.Port
	}

	username := cfg.Username
	if username == "" {
		if entry != nil && entry.User != nil {
			username = *entry.User
		} else {
			username = os.Getenv("USER")
			if username == "" {
				username = "root"
			}
		}
	}

	// Build known-hosts callback (security boundary — never InsecureIgnoreHostKey).
	khPath, err := DefaultKnownHostsPath()
	if err != nil {
		return nil, fmt.Errorf("known_hosts path: %w", err)
	}
	hostKeyCallback := HostKeyCallbackFromFile(khPath)

	// Build auth methods.
	authMethods, err := buildAuthMethods(cfg, entry)
	if err != nil {
		return nil, err
	}

	clientCfg := &gossh.ClientConfig{
		User:            username,
		Auth:            authMethods,
		HostKeyCallback: hostKeyCallback,
		Timeout:         120 * time.Second,
	}

	addr := fmt.Sprintf("%s:%d", actualHost, actualPort)
	tcpConn, err := dialer.Dial("tcp", addr)
	if err != nil {
		return nil, fmt.Errorf("SSH connection failed: %w", err)
	}

	sshConn, chans, reqs, err := gossh.NewClientConn(tcpConn, addr, clientCfg)
	if err != nil {
		_ = tcpConn.Close()
		return nil, fmt.Errorf("SSH connection failed: %w", err)
	}
	client := gossh.NewClient(sshConn, chans, reqs)

	// Open SFTP subsystem.
	sftpClient, err := sftp.NewClient(client)
	if err != nil {
		_ = client.Close()
		return nil, fmt.Errorf("failed to start SFTP session: %w", err)
	}

	// Resolve remote paths.
	remoteHome := resolveRemoteHome(client)
	remoteProjectsPath := findRemoteProjectsPath(sftpClient, username, remoteHome)
	remoteTodosPath := ""
	if remoteHome != "" {
		remoteTodosPath = remoteHome + "/.claude/todos"
	} else {
		remoteTodosPath = "/home/" + username + "/.claude/todos"
	}

	return &Connection{
		client:             client,
		sftp:               sftpClient,
		RemoteProjectsPath: remoteProjectsPath,
		remoteTodosPath:    remoteTodosPath,
		host:               cfg.Host,
	}, nil
}

// TestConnection mirrors test_connection — connect then immediately drop.
func TestConnection(cfg *ConnectionConfig, dialer Dialer) error {
	conn, err := Connect(cfg, dialer)
	if err != nil {
		return err
	}
	conn.close()
	return nil
}

// ConnectWithRetry mirrors connect_with_retry — exponential back-off on transient errors.
// onRetry is called before each sleep so the caller can emit status events.
// NEVER called while holding s.mu.
func ConnectWithRetry(
	cfg *ConnectionConfig,
	retryCfg RetryConfig,
	dialer Dialer,
	onRetry func(attempt, maxRetries uint32, errMsg string),
) (*Connection, error) {
	state := RetryState{}
	for {
		conn, err := Connect(cfg, dialer)
		if err == nil {
			return conn, nil
		}
		if !IsTransientError(err.Error()) || !state.CanRetry(retryCfg) {
			return nil, err
		}
		delay := state.NextDelay(retryCfg)
		state.Advance(err.Error())
		if onRetry != nil {
			onRetry(state.Attempt, retryCfg.MaxRetries, err.Error())
		}
		time.Sleep(delay)
	}
}

// buildAuthMethods constructs the []gossh.AuthMethod slice.
// Mirrors the authenticate() function in the Rust source.
func buildAuthMethods(cfg *ConnectionConfig, entry *ConfigHostEntry) ([]gossh.AuthMethod, error) {
	switch cfg.AuthMethod {
	case "password":
		if cfg.Password == nil {
			return nil, fmt.Errorf("password required")
		}
		return []gossh.AuthMethod{gossh.Password(*cfg.Password)}, nil

	case "privateKey":
		keyPath := "~/.ssh/id_rsa"
		if cfg.PrivateKeyPath != nil {
			keyPath = *cfg.PrivateKeyPath
		}
		signer, err := loadPrivateKey(expandTilde(keyPath))
		if err != nil {
			return nil, fmt.Errorf("cannot read private key at %s: %w", keyPath, err)
		}
		return []gossh.AuthMethod{gossh.PublicKeys(signer)}, nil

	case "agent":
		sock := DiscoverAgentSocket()
		if sock == "" {
			return nil, fmt.Errorf("SSH agent socket not found")
		}
		conn, err := net.Dial("unix", sock)
		if err != nil {
			return nil, fmt.Errorf("cannot connect to SSH agent: %w", err)
		}
		ag := agent.NewClient(conn)
		return []gossh.AuthMethod{gossh.PublicKeysCallback(ag.Signers)}, nil

	default:
		// "auto" — try config identity file, then agent, then default key files.
		var methods []gossh.AuthMethod

		if entry != nil && entry.HasIdentityFile {
			for _, name := range []string{"id_ed25519", "id_rsa", "id_ecdsa"} {
				home, _ := os.UserHomeDir()
				if signer, err := loadPrivateKey(filepath.Join(home, ".ssh", name)); err == nil {
					methods = append(methods, gossh.PublicKeys(signer))
				}
			}
		}

		if sock := DiscoverAgentSocket(); sock != "" {
			if conn, err := net.Dial("unix", sock); err == nil {
				ag := agent.NewClient(conn)
				methods = append(methods, gossh.PublicKeysCallback(ag.Signers))
			}
		}

		for _, name := range []string{"id_ed25519", "id_rsa", "id_ecdsa"} {
			home, _ := os.UserHomeDir()
			if signer, err := loadPrivateKey(filepath.Join(home, ".ssh", name)); err == nil {
				methods = append(methods, gossh.PublicKeys(signer))
			}
		}

		if len(methods) == 0 {
			return nil, fmt.Errorf("no authentication method succeeded")
		}
		return methods, nil
	}
}

// loadPrivateKey reads and parses a PEM private key file (no passphrase).
func loadPrivateKey(path string) (gossh.Signer, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return gossh.ParsePrivateKey(data)
}

// expandTilde replaces a leading ~ with the home directory.
func expandTilde(path string) string {
	if !strings.HasPrefix(path, "~") {
		return path
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return path
	}
	return home + path[1:]
}

// resolveRemoteHome mirrors resolve_remote_home — runs `printf %s "$HOME"` via exec.
func resolveRemoteHome(client *gossh.Client) string {
	sess, err := client.NewSession()
	if err != nil {
		return ""
	}
	defer sess.Close()
	out, err := sess.Output(`printf %s "$HOME"`)
	if err != nil {
		return ""
	}
	trimmed := strings.TrimSpace(string(out))
	if strings.HasPrefix(trimmed, "/") {
		return trimmed
	}
	return ""
}

// findRemoteProjectsPath mirrors find_remote_projects_path — checks candidates via SFTP Stat.
func findRemoteProjectsPath(sftpClient *sftp.Client, username, remoteHome string) string {
	var candidates []string
	if remoteHome != "" {
		candidates = append(candidates, remoteHome+"/.claude/projects")
	}
	// Deduplicate via seen set (map iteration randomized — sort not needed here
	// since order matters for preference, not for correctness).
	seen := map[string]bool{}
	for _, c := range []string{
		"/home/" + username + "/.claude/projects",
		"/Users/" + username + "/.claude/projects",
		"/root/.claude/projects",
	} {
		if !seen[c] {
			seen[c] = true
			candidates = append(candidates, c)
		}
	}

	for _, c := range candidates {
		if _, err := sftpClient.Stat(c); err == nil {
			return c
		}
	}

	if remoteHome != "" {
		return remoteHome + "/.claude/projects"
	}
	return "/home/" + username + "/.claude/projects"
}
